intake/readers/readers.py

```diff
@@ -513,36 +513,117 @@ def _read(self, name, **kw):
 class LlamaServerReader(BaseReader):
     """Create llama.cpp server using local pretrained model file"""
 
-    output_instance = "intake.readers.datatypes:ChatService"
-    implements = {datatypes.GGUF, datatypes.SafeTensors}
-    imports = {"transformers"}
+    output_instance = "intake.readers.datatypes:LlamaCPPService"
+    implements = {datatypes.GGUF}
+    imports = {"requests"}
+
+    @classmethod
+    def _find_executable(cls):
+        import shutil
+
+        # executables were renamed in https://github.com/ggerganov/llama.cpp/pull/7809
+        path = shutil.which("llama-server")
+        if path is None:
+            # fallback on old name
+            path = shutil.which("server")
+        return path
+
+    @classmethod
+    def check_imports(cls):
+        imports = super().check_imports()
+        path = cls._find_executable()
+        return imports & (path is not None)
+
+    def _local_model_path(self, data, callback=None):
+        import os
+        from fsspec.core import split_protocol
+        from intake.catalog.default import user_data_dir
+
+        if callback is None:
+            from fsspec.callbacks import _DEFAULT_CALLBACK
+
+            cb = _DEFAULT_CALLBACK
+        else:
+            cb = callback("model")
+
+        protocol, _ = split_protocol(data.url)
+        if protocol is None:
+            # no protocol means local path
+            return data.url
+
+        storage_options = {} if data.storage_options is None else data.storage_options
+        cache_location = os.path.join(user_data_dir(), "llama.cpp")
+        options = {protocol: storage_options, "simplecache": {"cache_storage": cache_location}}
+        fs, path = fsspec.url_to_fs(f"simplecache::{data.url}", **options)
+
+        cached_fn = fs._check_file(path)
+        if cached_fn:
+            return cached_fn
+
+        with fs.fs.open(path) as f:
```

Comment:
TODO: fsspec caching filesystems should allow the setting of a callback for when they download. I believe this was suggested in the fsspec repo.

Reply:
Was there a reason to use `open/read` rather than `get()`

Reply:
What I've found is that while cachingfs methods supported a callback it was only activated when reading the file from the cache and not forwarded to the target fs while populating the cache. I'll get some time to make some PRs.

I also noticed some oddities with .get() where it would deadlock that I need to follow through on.

Reply:
I thought I had replied to this: probably fsspec is trying to list directories to find matches for the "?" in the URL, which looks like a glob wildcard. This can be suppressed by passing `get([path])`.

Reply:
`.get([path], [output_path], callback=...)` does avoid the deadlock, but the callback is on the list of files rather than a progress bar per file. I'll go with `.get_file()` for this case.

Reply:
At least for callbacks that implement `.branch()`